File: zig/src/core/lm.zig
const std = @import("std");

pub const ModelType = enum(u32) {
    mle = 0,
    lidstone = 1,
    kneser_ney_interpolated = 2,
};

const Counts = struct {
    unigram: std.AutoHashMapUnmanaged(u32, u32) = .empty,
    bigram: std.AutoHashMapUnmanaged(u64, u32) = .empty,
    trigram: std.AutoHashMapUnmanaged(u128, u32) = .empty,

    followers1: std.AutoHashMapUnmanaged(u32, u32) = .empty,
    followers2: std.AutoHashMapUnmanaged(u64, u32) = .empty,
    continuation: std.AutoHashMapUnmanaged(u32, u32) = .empty,

    seen_bigram: std.AutoHashMapUnmanaged(u64, void) = .empty,
    seen_trigram: std.AutoHashMapUnmanaged(u128, void) = .empty,

    unigram_total: u64 = 0,
    continuation_type_count: u64 = 0,

    fn deinit(self: *Counts, allocator: std.mem.Allocator) void {
        self.unigram.deinit(allocator);
        self.bigram.deinit(allocator);
        self.trigram.deinit(allocator);
        self.followers1.deinit(allocator);
        self.followers2.deinit(allocator);
        self.continuation.deinit(allocator);
        self.seen_bigram.deinit(allocator);
        self.seen_trigram.deinit(allocator);
    }
};

fn incU32(map: *std.AutoHashMapUnmanaged(u32, u32), allocator: std.mem.Allocator, k: u32) !void {
    const entry = try map.getOrPut(allocator, k);
    if (!entry.found_existing) entry.value_ptr.* = 0;
    entry.value_ptr.* += 1;
}

fn incU64(map: *std.AutoHashMapUnmanaged(u64, u32), allocator: std.mem.Allocator, k: u64) !void {
    const entry = try map.getOrPut(allocator, k);
    if (!entry.found_existing) entry.value_ptr.* = 0;
    entry.value_ptr.* += 1;
}

fn incU128(map: *std.AutoHashMapUnmanaged(u128, u32), allocator: std.mem.Allocator, k: u128) !void {
    const entry = try map.getOrPut(allocator, k);
    if (!entry.found_existing) entry.value_ptr.* = 0;
    entry.value_ptr.* += 1;
}

fn keyBigram(a: u32, b: u32) u64 {
    return (@as(u64, a) << 32) | @as(u64, b);
}

fn keyTrigram(a: u32, b: u32, c: u32) u128 {
    return (@as(u128, a) << 64) | (@as(u128, b) << 32) | @as(u128, c);
}

fn markSeenBigram(counts: *Counts, allocator: std.mem.Allocator, prev: u32, word: u32) !void {
    const k = keyBigram(prev, word);
    const entry = try counts.seen_bigram.getOrPut(allocator, k);
    if (!entry.found_existing) {
        const follower = try counts.followers1.getOrPut(allocator, prev);
        if (!follower.found_existing) follower.value_ptr.* = 0;
        follower.value_ptr.* += 1;

        const cont = try counts.continuation.getOrPut(allocator, word);
        if (!cont.found_existing) cont.value_ptr.* = 0;
        cont.value_ptr.* += 1;
    }
}

fn markSeenTrigram(counts: *Counts, allocator: std.mem.Allocator, a: u32, b: u32, c: u32) !void {
    const trigram_k = keyTrigram(a, b, c);
    const entry = try counts.seen_trigram.getOrPut(allocator, trigram_k);
    if (!entry.found_existing) {
        const context_k = keyBigram(a, b);
        const follower = try counts.followers2.getOrPut(allocator, context_k);
        if (!follower.found_existing) follower.value_ptr.* = 0;
        follower.value_ptr.* += 1;
    }
}

fn buildCounts(
    token_ids: []const u32,
    sentence_offsets: []const u32,
    order: u32,
    allocator: std.mem.Allocator,
) !Counts {
    var counts: Counts = .{};
    errdefer counts.deinit(allocator);

    var sent_idx: usize = 0;
    while (sent_idx + 1 < sentence_offsets.len) : (sent_idx += 1) {
        const start = @as(usize, sentence_offsets[sent_idx]);
        const end = @as(usize, sentence_offsets[sent_idx + 1]);
        if (end <= start or end > token_ids.len) continue;
        const sentence = token_ids[start..end];

        for (sentence, 0..) |tok, i| {
            try incU32(&counts.unigram, allocator, tok);
            counts.unigram_total += 1;

            if (order >= 2 and i >= 1) {
                const prev = sentence[i - 1];
                try incU64(&counts.bigram, allocator, keyBigram(prev, tok));
                try markSeenBigram(&counts, allocator, prev, tok);
            }
            if (order >= 3 and i >= 2) {
                const a = sentence[i - 2];
                const b = sentence[i - 1];
                try incU128(&counts.trigram, allocator, keyTrigram(a, b, tok));
                try markSeenTrigram(&counts, allocator, a, b, tok);
            }
        }
    }

    counts.continuation_type_count = counts.seen_bigram.count();
    return counts;
}

fn countContext(counts: *const Counts, ctx: []const u32) u32 {
    if (ctx.len == 0) {
        if (counts.unigram_total > std.math.maxInt(u32)) return std.math.maxInt(u32);
        return @intCast(counts.unigram_total);
    }
    if (ctx.len == 1) {
        return counts.unigram.get(ctx[0]) orelse 0;
    }
    if (ctx.len == 2) {
        return counts.bigram.get(keyBigram(ctx[0], ctx[1])) orelse 0;
    }
    return 0;
}

fn ngramCount(counts: *const Counts, ctx: []const u32, word: u32) u32 {
    if (ctx.len == 0) return counts.unigram.get(word) orelse 0;
    if (ctx.len == 1) return counts.bigram.get(keyBigram(ctx[0], word)) orelse 0;
    if (ctx.len == 2) return counts.trigram.get(keyTrigram(ctx[0], ctx[1], word)) orelse 0;
    return 0;
}

fn followerCount(counts: *const Counts, ctx: []const u32) u32 {
    if (ctx.len == 1) return counts.followers1.get(ctx[0]) orelse 0;
    if (ctx.len == 2) return counts.followers2.get(keyBigram(ctx[0], ctx[1])) orelse 0;
    return 0;
}

fn backoffTail(ctx: []const u32) []const u32 {
    if (ctx.len == 0) return ctx;
    return ctx[1..];
}

fn scoreMle(counts: *const Counts, word: u32, ctx: []const u32) f64 {
    if (ctx.len == 0) {
        if (counts.unigram_total == 0) return 0;
        return @as(f64, @floatFromInt(ngramCount(counts, ctx, word))) / @as(f64, @floatFromInt(counts.unigram_total));
    }
    const ctx_count = countContext(counts, ctx);
    if (ctx_count == 0) return scoreMle(counts, word, backoffTail(ctx));
    const gram_count = ngramCount(counts, ctx, word);
    return @as(f64, @floatFromInt(gram_count)) / @as(f64, @floatFromInt(ctx_count));
}

fn scoreLidstone(counts: *const Counts, word: u32, ctx: []const u32, gamma: f64, vocab_size: u32) f64 {
    const vocab_f = @as(f64, @floatFromInt(@max(@as(u32, 1), vocab_size)));
    if (ctx.len == 0) {
        const gram = @as(f64, @floatFromInt(ngramCount(counts, ctx, word)));
        const denom = @as(f64, @floatFromInt(counts.unigram_total)) + gamma * vocab_f;
        if (denom <= 0) return 0;
        return (gram + gamma) / denom;
    }
    const ctx_count = countContext(counts, ctx);
    if (ctx_count == 0) return scoreLidstone(counts, word, backoffTail(ctx), gamma, vocab_size);
    const gram_count = @as(f64, @floatFromInt(ngramCount(counts, ctx, word)));
    const denom = @as(f64, @floatFromInt(ctx_count)) + gamma * vocab_f;
    if (denom <= 0) return 0;
    return (gram_count + gamma) / denom;
}

fn continuationProb(counts: *const Counts, word: u32) f64 {
    if (counts.continuation_type_count == 0) return 0;
    const cont = counts.continuation.get(word) orelse 0;
    if (cont == 0) {
        return 1.0 / (@as(f64, @floatFromInt(counts.continuation_type_count)) * 10.0);
    }
    return @as(f64, @floatFromInt(cont)) / @as(f64, @floatFromInt(counts.continuation_type_count));
}

fn scoreKneserNey(counts: *const Counts, word: u32, ctx: []const u32, discount: f64) f64 {
    if (ctx.len == 0) return continuationProb(counts, word);

    const ctx_count = countContext(counts, ctx);
    if (ctx_count == 0) return scoreKneserNey(counts, word, backoffTail(ctx), discount);

    const gram_count = ngramCount(counts, ctx, word);
    const followers = followerCount(counts, ctx);

    const ctx_f = @as(f64, @floatFromInt(ctx_count));
    const discounted = @max(@as(f64, @floatFromInt(gram_count)) - discount, 0.0) / ctx_f;
    const lambda = (discount * @as(f64, @floatFromInt(followers))) / ctx_f;
    return discounted + lambda * scoreKneserNey(counts, word, backoffTail(ctx), discount);
}

fn scoreWord(
    counts: *const Counts,
    word: u32,
    context: []const u32,
    order: u32,
    model: ModelType,
    gamma: f64,
    discount: f64,
    vocab_size: u32,
) f64 {
    const keep = if (context.len > @as(usize, order -| 1)) context[context.len - @as(usize, order -| 1) ..] else context;
    return switch (model) {
        .mle => scoreMle(counts, word, keep),
        .lidstone => scoreLidstone(counts, word, keep, gamma, vocab_size),
        .kneser_ney_interpolated => scoreKneserNey(counts, word, keep, discount),
    };
}

pub fn evalIds(
    token_ids: []const u32,
    sentence_offsets: []const u32,
    order: u32,
    model: ModelType,
    gamma: f64,
    discount: f64,
    vocab_size: u32,
    probe_context_flat: []const u32,
    probe_context_lens: []const u32,
    probe_words: []const u32,
    out_scores: []f64,
    perplexity_tokens: []const u32,
    prefix_tokens: []const u32,
    allocator: std.mem.Allocator,
) !f64 {
    var counts = try buildCounts(token_ids, sentence_offsets, order, allocator);
    defer counts.deinit(allocator);

    var ctx_cursor: usize = 0;
    const probe_count = @min(@min(probe_context_lens.len, probe_words.len), out_scores.len);
    var i: usize = 0;
    while (i < probe_count) : (i += 1) {
        const ctx_len = @as(usize, probe_context_lens[i]);
        if (ctx_cursor + ctx_len > probe_context_flat.len) {
            out_scores[i] = 0;
            continue;
        }
        const ctx = probe_context_flat[ctx_cursor .. ctx_cursor + ctx_len];
        ctx_cursor += ctx_len;
        out_scores[i] = scoreWord(&counts, probe_words[i], ctx, order, model, gamma, discount, vocab_size);
    }

    if (perplexity_tokens.len == 0) return std.math.inf(f64);
    var history = std.ArrayListUnmanaged(u32).empty;
    defer history.deinit(allocator);
    try history.appendSlice(allocator, prefix_tokens);

    var neg_log2: f64 = 0;
    for (perplexity_tokens) |tok| {
        const keep_len = @min(history.items.len, @as(usize, order -| 1));
        const ctx = if (keep_len == 0) history.items[0..0] else history.items[history.items.len - keep_len ..];
        var prob = scoreWord(&counts, tok, ctx, order, model, gamma, discount, vocab_size);
        if (!std.math.isFinite(prob) or prob <= 0) prob = 1e-12;
        neg_log2 += -std.math.log2(prob);
        try history.append(allocator, tok);
    }

    return std.math.pow(f64, 2.0, neg_log2 / @as(f64, @floatFromInt(perplexity_tokens.len)));
}

test "lm eval ids basic parity sanity" {
    const allocator = std.testing.allocator;
    const tokens = [_]u32{ 1, 2, 3, 4, 1, 2, 5, 4 };
    const offsets = [_]u32{ 0, 4, 8 };
    const probe_ctx = [_]u32{ 1, 2, 1, 2 };
    const probe_lens = [_]u32{ 2, 2 };
    const probe_words = [_]u32{ 3, 5 };
    var out = [_]f64{0} ** 2;
    const perplexity_tokens = [_]u32{ 1, 2, 3, 4 };
    const prefix = [_]u32{ 0, 0 };

    const ppl = try evalIds(
        &tokens,
        &offsets,
        3,
        .kneser_ney_interpolated,
        0.1,
        0.75,
        6,
        &probe_ctx,
        &probe_lens,
        &probe_words,
        &out,
        &perplexity_tokens,
        &prefix,
        allocator,
    );
    try std.testing.expect(out[0] > 0);
    try std.testing.expect(out[1] > 0);
    try std.testing.expect(std.math.isFinite(ppl));
}

File: rust/src/lm.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ModelType {
    Mle = 0,
    Lidstone = 1,
    KneserNeyInterpolated = 2,
}

struct Counts {
    unigram: HashMap<u32, u32>,
    bigram: HashMap<u64, u32>,
    trigram: HashMap<u128, u32>,

    followers1: HashMap<u32, u32>,
    followers2: HashMap<u64, u32>,
    continuation: HashMap<u32, u32>,

    seen_bigram_len: u64,
    unigram_total: u64,
    continuation_type_count: u64,
}

fn key_bigram(a: u32, b: u32) -> u64 {
    ((a as u64) << 32) | (b as u64)
}

fn key_trigram(a: u32, b: u32, c: u32) -> u128 {
    ((a as u128) << 64) | ((b as u128) << 32) | (c as u128)
}

fn build_counts(token_ids: &[u32], sentence_offsets: &[u32], order: u32) -> Counts {
    let mut counts = Counts {
        unigram: HashMap::new(),
        bigram: HashMap::new(),
        trigram: HashMap::new(),
        followers1: HashMap::new(),
        followers2: HashMap::new(),
        continuation: HashMap::new(),
        seen_bigram_len: 0,
        unigram_total: 0,
        continuation_type_count: 0,
    };

    // Dedup sets are only needed during construction; we track sizes for
    // continuation_type_count and use the follower/continuation maps' insert
    // semantics to mirror "first time this bigram is seen".
    let mut seen_bigram: std::collections::HashSet<u64> = std::collections::HashSet::new();
    let mut seen_trigram: std::collections::HashSet<u128> = std::collections::HashSet::new();

    let mut sent_idx: usize = 0;
    while sent_idx + 1 < sentence_offsets.len() {
        let start = sentence_offsets[sent_idx] as usize;
        let end = sentence_offsets[sent_idx + 1] as usize;
        sent_idx += 1;
        if end <= start || end > token_ids.len() {
            continue;
        }
        let sentence = &token_ids[start..end];

        for (i, &tok) in sentence.iter().enumerate() {
            *counts.unigram.entry(tok).or_insert(0) += 1;
            counts.unigram_total += 1;

            if order >= 2 && i >= 1 {
                let prev = sentence[i - 1];
                *counts.bigram.entry(key_bigram(prev, tok)).or_insert(0) += 1;

                let k = key_bigram(prev, tok);
                if seen_bigram.insert(k) {
                    *counts.followers1.entry(prev).or_insert(0) += 1;
                    *counts.continuation.entry(tok).or_insert(0) += 1;
                }
            }
            if order >= 3 && i >= 2 {
                let a = sentence[i - 2];
                let b = sentence[i - 1];
                *counts.trigram.entry(key_trigram(a, b, tok)).or_insert(0) += 1;

                let trigram_k = key_trigram(a, b, tok);
                if seen_trigram.insert(trigram_k) {
                    let context_k = key_bigram(a, b);
                    *counts.followers2.entry(context_k).or_insert(0) += 1;
                }
            }
        }
    }

    counts.seen_bigram_len = seen_bigram.len() as u64;
    counts.continuation_type_count = counts.seen_bigram_len;
    counts
}

fn count_context(counts: &Counts, ctx: &[u32]) -> u32 {
    match ctx.len() {
        0 => {
            if counts.unigram_total > u32::MAX as u64 {
                u32::MAX
            } else {
                counts.unigram_total as u32
            }
        }
        1 => counts.unigram.get(&ctx[0]).copied().unwrap_or(0),
        2 => counts.bigram.get(&key_bigram(ctx[0], ctx[1])).copied().unwrap_or(0),
        _ => 0,
    }
}

fn ngram_count(counts: &Counts, ctx: &[u32], word: u32) -> u32 {
    match ctx.len() {
        0 => counts.unigram.get(&word).copied().unwrap_or(0),
        1 => counts.bigram.get(&key_bigram(ctx[0], word)).copied().unwrap_or(0),
        2 => counts
            .trigram
            .get(&key_trigram(ctx[0], ctx[1], word))
            .copied()
            .unwrap_or(0),
        _ => 0,
    }
}

fn follower_count(counts: &Counts, ctx: &[u32]) -> u32 {
    match ctx.len() {
        1 => counts.followers1.get(&ctx[0]).copied().unwrap_or(0),
        2 => counts.followers2.get(&key_bigram(ctx[0], ctx[1])).copied().unwrap_or(0),
        _ => 0,
    }
}

fn backoff_tail(ctx: &[u32]) -> &[u32] {
    if ctx.is_empty() {
        return ctx;
    }
    &ctx[1..]
}

fn score_mle(counts: &Counts, word: u32, ctx: &[u32]) -> f64 {
    if ctx.is_empty() {
        if counts.unigram_total == 0 {
            return 0.0;
        }
        return (ngram_count(counts, ctx, word) as f64) / (counts.unigram_total as f64);
    }
    let ctx_count = count_context(counts, ctx);
    if ctx_count == 0 {
        return score_mle(counts, word, backoff_tail(ctx));
    }
    let gram_count = ngram_count(counts, ctx, word);
    (gram_count as f64) / (ctx_count as f64)
}

fn score_lidstone(counts: &Counts, word: u32, ctx: &[u32], gamma: f64, vocab_size: u32) -> f64 {
    let vocab_f = 1u32.max(vocab_size) as f64;
    if ctx.is_empty() {
        let gram = ngram_count(counts, ctx, word) as f64;
        let denom = (counts.unigram_total as f64) + gamma * vocab_f;
        if denom <= 0.0 {
            return 0.0;
        }
        return (gram + gamma) / denom;
    }
    let ctx_count = count_context(counts, ctx);
    if ctx_count == 0 {
        return score_lidstone(counts, word, backoff_tail(ctx), gamma, vocab_size);
    }
    let gram_count = ngram_count(counts, ctx, word) as f64;
    let denom = (ctx_count as f64) + gamma * vocab_f;
    if denom <= 0.0 {
        return 0.0;
    }
    (gram_count + gamma) / denom
}

fn continuation_prob(counts: &Counts, word: u32) -> f64 {
    if counts.continuation_type_count == 0 {
        return 0.0;
    }
    let cont = counts.continuation.get(&word).copied().unwrap_or(0);
    if cont == 0 {
        return 1.0 / ((counts.continuation_type_count as f64) * 10.0);
    }
    (cont as f64) / (counts.continuation_type_count as f64)
}

fn score_kneser_ney(counts: &Counts, word: u32, ctx: &[u32], discount: f64) -> f64 {
    if ctx.is_empty() {
        return continuation_prob(counts, word);
    }

    let ctx_count = count_context(counts, ctx);
    if ctx_count == 0 {
        return score_kneser_ney(counts, word, backoff_tail(ctx), discount);
    }

    let gram_count = ngram_count(counts, ctx, word);
    let followers = follower_count(counts, ctx);

    let ctx_f = ctx_count as f64;
    let discounted = (((gram_count as f64) - discount).max(0.0)) / ctx_f;
    let lambda = (discount * (followers as f64)) / ctx_f;
    discounted + lambda * score_kneser_ney(counts, word, backoff_tail(ctx), discount)
}

fn score_word(
    counts: &Counts,
    word: u32,
    context: &[u32],
    order: u32,
    model: ModelType,
    gamma: f64,
    discount: f64,
    vocab_size: u32,
) -> f64 {
    let keep_len = (order.saturating_sub(1)) as usize;
    let keep: &[u32] = if context.len() > keep_len {
        &context[context.len() - keep_len..]
    } else {
        context
    };
    match model {
        ModelType::Mle => score_mle(counts, word, keep),
        ModelType::Lidstone => score_lidstone(counts, word, keep, gamma, vocab_size),
        ModelType::KneserNeyInterpolated => score_kneser_ney(counts, word, keep, discount),
    }
}

pub fn eval_ids(
    token_ids: &[u32],
    sentence_offsets: &[u32],
    order: u32,
    model: ModelType,
    gamma: f64,
    discount: f64,
    vocab_size: u32,
    probe_context_flat: &[u32],
    probe_context_lens: &[u32],
    probe_words: &[u32],
    out_scores: &mut [f64],
    perplexity_tokens: &[u32],
    prefix_tokens: &[u32],
) -> f64 {
    let counts = build_counts(token_ids, sentence_offsets, order);

    let mut ctx_cursor: usize = 0;
    let probe_count = probe_context_lens
        .len()
        .min(probe_words.len())
        .min(out_scores.len());
    for i in 0..probe_count {
        let ctx_len = probe_context_lens[i] as usize;
        if ctx_cursor + ctx_len > probe_context_flat.len() {
            out_scores[i] = 0.0;
            continue;
        }
        let ctx = &probe_context_flat[ctx_cursor..ctx_cursor + ctx_len];
        ctx_cursor += ctx_len;
        out_scores[i] = score_word(
            &counts,
            probe_words[i],
            ctx,
            order,
            model,
            gamma,
            discount,
            vocab_size,
        );
    }

    if perplexity_tokens.is_empty() {
        return f64::INFINITY;
    }
    let mut history: Vec<u32> = prefix_tokens.to_vec();

    let mut neg_log2: f64 = 0.0;
    for &tok in perplexity_tokens {
        let keep_len = history.len().min((order.saturating_sub(1)) as usize);
        let ctx: &[u32] = if keep_len == 0 {
            &history[0..0]
        } else {
            &history[history.len() - keep_len..]
        };
        let mut prob = score_word(&counts, tok, ctx, order, model, gamma, discount, vocab_size);
        if !prob.is_finite() || prob <= 0.0 {
            prob = 1e-12;
        }
        neg_log2 += -prob.log2();
        history.push(tok);
    }

    2f64.powf(neg_log2 / (perplexity_tokens.len() as f64))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lm_eval_ids_basic_parity_sanity() {
        let tokens: [u32; 8] = [1, 2, 3, 4, 1, 2, 5, 4];
        let offsets: [u32; 3] = [0, 4, 8];
        let probe_ctx: [u32; 4] = [1, 2, 1, 2];
        let probe_lens: [u32; 2] = [2, 2];
        let probe_words: [u32; 2] = [3, 5];
        let mut out = [0f64; 2];
        let perplexity_tokens: [u32; 4] = [1, 2, 3, 4];
        let prefix: [u32; 2] = [0, 0];

        let ppl = eval_ids(
            &tokens,
            &offsets,
            3,
            ModelType::KneserNeyInterpolated,
            0.1,
            0.75,
            6,
            &probe_ctx,
            &probe_lens,
            &probe_words,
            &mut out,
            &perplexity_tokens,
            &prefix,
        );
        assert!(out[0] > 0.0);
        assert!(out[1] > 0.0);
        assert!(ppl.is_finite());
    }
}
